src/rules/unnecessary_function_arguments.erl

```diff
@@ -3,24 +3,111 @@
         Node <- AST,
         erl_syntax:type(Node) == function,
+        is_not_a_callback(ImpCallbacks, File, Node),
         Result <- analyze_function(File, Node)].
 
+%% @doc Constructs a map with the callbacks of all the files.
+%% 1. collect all the behaviors that the file implements.
+%% 2. for each one of them evaluate with BehaviourMod:behaviour_info(callbacks)
+%%    and keep the tuples in a single list.
+%%    2.1. if that call fails, look for the module/s in all the files.
+%%    2.2. if the module is not there, ignore the whole file
+%%         (we can't be sure which functions are callbacks).
+%%    2.3. If the module is there collect all callback attributes and add them
+%%         to the list obtained in the item 2.
+-spec callback_usage(hank_rule:asts()) -> imp_callbacks().
+callback_usage(FilesAndASTs) ->
+    lists:foldl(fun({File, AST}, Result) ->
+                   FoldFun =
+                       fun(Node, FileCallbacks) ->
+                          case hank_utils:node_has_attrs(Node, [behaviour, behavior]) of
+                              true ->
+                                  {_, _, _, {attribute, _, [{_, _, _, BehaviourMod}]}} = Node,
+                                  Callbacks =
+                                      try
+                                          %% For pre-compiled behaviours like OTP ones
+                                          BehaviourMod:behaviour_info(callbacks)
+                                      catch
+                                          %% Local behaviours will raise an exception
+                                          _:_ ->
+                                              case not_compiled_behaviour_callbacks(BehaviourMod,
+                                                                                    FilesAndASTs)
+                                              of
+                                                  ignore ->
+                                                      %% This is a way to break the foldl.
+                                                      %% If some behaviour cannot be readed,
+                                                      %% ignore the whole file
+                                                      throw(ignore);
+                                                  Callbacks0 ->
+                                                      Callbacks0
+                                              end
+                                      end,
+                                  FileCallbacks ++ Callbacks;
+                              _ ->
+                                  FileCallbacks
+                          end
+                       end,
+                   ResultsForFile =
+                       try
+                           erl_syntax_lib:fold(FoldFun, [], erl_syntax:form_list(AST))
+                       catch
+                           ignore ->
+                               ignore
```

Comment:
I wonder if returning `[]` here would not make the code simpler… Can we have a behavior with no callbacks? I don't think so.

Reply:
It's different `[]` which means that the file should be analyzed but does not contain callbacks than `ignore` which means _ignore_ this file.

Reply:
OK. I see.